### PR TITLE
feat: add customised, tracked method for metadata caching

### LINKED ISSUE #70: Move from printing to logging
We should consider how to make logging a bit more friendly for the full range of users. We could move to using python logging instead of printing in the main functions, and then allow users to silence some of their output, though python logging doesn't make this ultra friendly. We could have some other setting that we keep (I don't love adding more to the state of the class, but maybe that's reasonable?). Other options are welcome!

### PR DESCRIPTION
**Pull request by Soap2G**
This PR introduces 2 main new features:
1. streaming of progress while caching the metadata of a release
2. possibility to customise the page size, to tolerate possible network errors

Please notice that [this commit on our APIs](https://gitlab.cern.ch/gguerrie/atlasopenmagic-rest-api/-/commit/491fa4e0069b68feb799f3172844c8299d7be443) introduces a few new items, such as:

1. Add /datasets/count endpoint with release filtering
2. Add /datasets/lean for efficient paginated listings
3. Enable gzip compression for large responses

### More goodies
In addition, I've added these items in order to make the whole thing a bit more robust:

#### Logging
Switched to python logging, with possibility to configure verbosity.
Closes #70 

#### Session Management
  - HTTP connection pooling (because why not DoS our APIs?)
  - Automatic retry logic (5 total retries with exponential backoff)
  - Handles 429/5xx errors "gracefully" (it doesn't just break)
  - Persistent session with gzip compression (overkill, but we never know)

#### Error Handling
  - Defensive checks for empty API responses

#### Tests

  - Tests for edge cases: empty responses, cache corruption, race conditions
  - Mock support for pagination, count endpoints, and individual dataset lookups

### CODE REVIEW

File: src/atlasopenmagic/metadata.py [hunk 1]
```diff
@@ -415,38 +469,41 @@ def find_all_files(local_path: str, warnmissing: bool = False) -> None:
...
-    # Retrieve the full dataset dictionary from the cache.
     sample_data = _metadata.get(key_str)
 
+    # This check is still needed as a final safety net
     if not sample_data:
         raise ValueError(
-            f"Invalid key: '{key_str}'. \
-            No dataset found with this ID or name in release '{current_release}'."
+            f"Invalid key: '{key_str}'. "
+            f"No dataset found with this ID or name in release '{current_release}'."
```
Comment (zlmarshall) [inline src/atlasopenmagic/metadata.py]:
```suggestion
            f"No dataset found with this ID or name in release: '{current_release}'."
```

File: src/atlasopenmagic/metadata.py
Comment (zlmarshall) [inline src/atlasopenmagic/metadata.py]:
```suggestion
        page_size: The number of records to retrieve at a time.

```

File: src/atlasopenmagic/metadata.py [hunk 2]
```diff
@@ -415,38 +469,41 @@ def find_all_files(local_path: str, warnmissing: bool = False) -> None:
...
-        var: A specific metadata field to retrieve.
-            If None, the entire metadata dictionary is returned.
-
-    Returns:
-        The full info dictionary for the dataset, or the value of the single field
-        if 'var' was specified.
-
-    Raises:
-        ValueError: If the dataset key or the specified variable field is not found.
-    """
```
[the remark this answers is not in the document]
Comment (Soap2G) [reply src/atlasopenmagic/metadata.py]: The refactoring rage sometimes goes too far; re-added.

**Comment by Soap2G** [comment]
@zlmarshall and @alexander-held please have a look to 
- https://github.com/atlas-outreach-data-tools/atlasopenmagic/commit/caad3061f84491720bfe6a0d636f93a8e6507287 (caching)
- https://github.com/atlas-outreach-data-tools/atlasopenmagic/pull/69/commits/1fb26a0305ea229b7b77673b72d7c2db16ce852a (logging)

**Comment by zlmarshall** [comment]
Awesome, thanks! One minor suggestion.

**Review by zlmarshall — APPROVED** [review]
Good as far as I’m concerned! Shall we go like this?
